transactable.go

```diff
@@ -108,6 +108,49 @@ func (c *transactableConn) QuerySingleJSON(
 	return runQuery(ctx, c, "QuerySingleJSON", cmd, out, args)
 }
 
+func (c *transactableConn) granularFlow(
+	ctx context.Context,
+	q *gfQuery,
+) error {
+	var (
+		err    error
+		edbErr Error
+	)
+
+	for i := 1; true; i++ {
+		if errors.As(err, &edbErr) && edbErr.HasTag(ShouldReconnect) {
+			err = c.reconnect(ctx, true)
+			if err != nil {
+				goto Error
+			}
+		}
+
+		err = c.reconnectingConn.granularFlow(ctx, q)
+
+	Error:
+		// q is a read only query if the server sent zero headers when q was
+		// prepared. Read only queries are retryable.
+		headers, _ := c.getCachedHeaders(q)
+
+		if len(headers) == 0 &&
```

Comment:
This is incorrect.  The query "readonliness" is indicated by a specific header: https://github.com/edgedb/edgedb/blob/02f95c12b35f51322050e7f3043f2cd159ef2a97/edb/server/protocol/binary.pyx#L115

Reply:
The actual capability flags are defined here: https://github.com/edgedb/edgedb/blob/02f95c12b35f51322050e7f3043f2cd159ef2a97/edb/server/compiler/enums.py#L40-L46

Essentially, the absence of the CAPABILITIY header or its value of `0` is an indicator that the query has no side effects.

Reply:
Actually, the fixed server would always return the `CAPABILITY` header, so if it's absent, it's the old server which forgets to send the header and so we should assume it isn't safe to retry.

Reply:
changed